samza-core/src/main/java/org/apache/samza/table/retry/RetriableWriteFunction.java

```diff
@@ -0,0 +1,131 @@
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import net.jodah.failsafe.AsyncFailsafe;
+import net.jodah.failsafe.Failsafe;
+import net.jodah.failsafe.RetryPolicy;
+
+
+/**
+ * Wrapper for a {@link TableWriteFunction} instance to add common retry
+ * support with a {@link TableRetryPolicy}. This wrapper is created by
+ * {@link org.apache.samza.table.remote.RemoteTableProvider} when a retry
+ * policy is specified together with the {@link TableWriteFunction}.
+ *
+ * Actual retry mechanism is provided by the failsafe library. Retry is
+ * attempted in an async way with a {@link ScheduledExecutorService}.
+ *
+ * @param <K> the type of the key in this table
+ * @param <V> the type of the value in this table
+ */
+public class RetriableWriteFunction<K, V> implements TableWriteFunction<K, V> {
+  private final RetryPolicy retryPolicy;
+  private final TableWriteFunction<K, V> writeFn;
+  private final ScheduledExecutorService retryExecutor;
+
+  @VisibleForTesting
+  RetryMetrics retryMetrics;
+
+  public RetriableWriteFunction(TableRetryPolicy policy, TableWriteFunction<K, V> writeFn,
+      ScheduledExecutorService retryExecutor)  {
+    Preconditions.checkNotNull(policy);
+    Preconditions.checkNotNull(writeFn);
+    Preconditions.checkNotNull(retryExecutor);
+
+    this.writeFn = writeFn;
+    this.retryExecutor = retryExecutor;
+    this.retryPolicy = policy.toFailsafePolicy(writeFn::isRetriable);
+  }
+
+  @Override
+  public CompletableFuture<Void> putAsync(K key, V record) {
+    return failsafe()
+        .future(k -> writeFn.putAsync(key, record))
+        .exceptionally(e -> {
+            throw new SamzaException("Failed to get the record for " + key + " after retries.", e);
+          });
+  }
+
+  @Override
+  public CompletableFuture<Void> putAllAsync(Collection<Entry<K, V>> records) {
+    return failsafe()
+        .future(k -> writeFn.putAllAsync(records))
+        .exceptionally(e -> {
+            throw new SamzaException("Failed to put records after retries.", e);
+          });
+  }
+
+  @Override
+  public CompletableFuture<Void> deleteAsync(K key) {
+    return failsafe()
+        .future(k -> writeFn.deleteAsync(key))
+        .exceptionally(e -> {
+            throw new SamzaException("Failed to delete the record for " + key + " after retries.", e);
+          });
+  }
+
+  @Override
+  public CompletableFuture<Void> deleteAllAsync(Collection<K> keys) {
+    return failsafe()
+        .future(k -> writeFn.deleteAllAsync(keys))
+        .exceptionally(e -> {
+            throw new SamzaException("Failed to delete the records for " + keys + " after retries.", e);
+          });
+  }
+
+  @Override
+  public boolean isRetriable(Throwable exception) {
+    return writeFn.isRetriable(exception);
+  }
+
+  private AsyncFailsafe<?> failsafe() {
+    long startMs = System.currentTimeMillis();
+    return Failsafe.with(retryPolicy).with(retryExecutor)
```

Comment:
Is this part reusable or do we have to create a new instance for every request?

Reply:
We need the instance because the timestamp is unique per request and each failsafe object contains its own context.